GGSmtp/GGSmtp/DefaultDataResponder.cs

```diff
@@ -0,0 +1,153 @@
+﻿using GGSmtp.Properties;
+using MimeKit;
+using Newtonsoft.Json;
+using RestSharp;
+using RestSharp.Serializers;
+using Simple.MailServer.Logging;
+using Simple.MailServer.Smtp;
+using Simple.MailServer.Smtp.Config;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GGSmtp
+{
+    public class DefaultDataResponder : SmtpDataResponder
+    {
+        private IDictionary<ISmtpSessionInfo, Stream> mSessionsData;
+        private IMailServerLogger mLogger;
+        private Settings mSettings;
+
+        public DefaultDataResponder(ISmtpServerConfiguration config) :
+            base(config)
+        {
+            mSessionsData = new Dictionary<ISmtpSessionInfo, Stream>();
+            mLogger = MailServerLogger.Instance;
+            mSettings = Settings.Default;
+        }
+
+        public override SmtpResponse DataStart(ISmtpSessionInfo sessionInfo)
+        {
+            mLogger.Info(string.Format("Start receiving mail from: {0}", sessionInfo.MailFrom));
+
+            mSessionsData[sessionInfo] = new MemoryStream();
+
+            return SmtpResponses.DataStart;
+        }
+
+        /// <summary>
+        /// Writes "a line" of data to a stream from the SMTP message
+        /// </summary>
+        /// <param name="sessionInfo"></param>
+        /// <param name="lineBuf"></param>
+        /// <returns></returns>
+        public override SmtpResponse DataLine(ISmtpSessionInfo sessionInfo, byte[] lineBuf)
+        {
+            Stream currentStream = mSessionsData[sessionInfo];
+
+            currentStream.Seek(0, SeekOrigin.End);
+            currentStream.Write(lineBuf, 0, lineBuf.Length);
+
+            currentStream.WriteByte(13);
+            currentStream.WriteByte(10);
+
+            return SmtpResponses.None;
+        }
+
+        public override SmtpResponse DataEnd(ISmtpSessionInfo sessionInfo)
+        {
+            Stream currentStream = mSessionsData[sessionInfo];
+            long streamLength = currentStream.Length;
+            string successMessage = string.Format("{0} bytes received", streamLength);
+
+            // Write a success response to return, but don't return it yet
+            SmtpResponse response = SmtpResponses.OK.CloneAndChange(successMessage);
+
+            try
+            {
+                currentStream.Seek(0, SeekOrigin.Begin);
+
+                // Read the mail message from the stream
+                MimeMessage message = MimeMessage.Load(currentStream);
+
+                foreach (MimePart attatchment in message.Attachments)
+                {
+                    mLogger.Debug(string.Format("Got attachment from {0}, sending . . .", sessionInfo.MailFrom));
+
+                    bool success = SendAttachment(sessionInfo.MailFrom.MailAddress, attatchment);
```

Comment:
Send attachment can be turned to void and throw an exception on failure.
It's wrapped in a try-catch anyways, and there's code duplication dealing it's failure !resolve